external_server/checkers/status_checker.py

```diff
@@ -0,0 +1,143 @@
+            )
+        else:
+            if self._allow_counter_reset:
+                self._counter = status.messageCounter
+                self._allow_counter_reset = False
+
+            self._received.put((status.messageCounter, status))
+            if status.messageCounter == self._counter:
+                while not self._received.empty() and self._received.queue[0][0] == self._counter:
+                    self._remove_oldest_skipped_and_stop_its_timer()
+                    oldest_received = self._received.get()[1]
+                    self._checked.put(oldest_received)
+                    self._counter += 1
+            else:  # status counter is greater than expected - some statuses are missing
+                self._store_skipped_counter_values(status.messageCounter)
+
+    def get(self) -> _Status | None:
+        """Returns the next Status message in queue if it is available, otherwise `None`."""
+        status = self._checked.get_nowait() if not self._checked.empty() else None
+        return status
+
+    def allow_counter_reset(self) -> None:
+        """The next status check will reset the counter to the received status counter value."""
+        self._allow_counter_reset = True
+
+    def set_counter(self, counter: CounterValue) -> None:
+        """Set the counter to the given value and disallow the counter reset."""
+        if self._received.empty() and self._checked.empty():
+            self._counter = counter
+            self._allow_counter_reset = False
+
+    def reset(self) -> None:
+        """Clear all data stored in the checker and reset the counter to the initial value.
+
+        Stop timers for all skipped status counter values.
+        """
+        self._clear_queues_and_timeout_event()
+        self._counter = StatusChecker.DEFAULT_INIT_COUNTER
+
+    def _clear_queues_and_timeout_event(self) -> None:
+        """Empty queues for received, skipped and checked statuses. Unset timeout event if set."""
+        self._clear_skipped_counters()
+        self._received.queue.clear()
+        self._checked.queue.clear()
+        self._timeout_event.clear()
+
+    def _clear_skipped_counters(self) -> None:
+        """Clear all skipped statuses and stop all timers."""
+        while not self._skipped.empty():
+            _, timer = self._skipped.get()
+            timer.cancel()
+            timer.join()
+
+    def _remove_oldest_skipped_and_stop_its_timer(self) -> None:
+        """Clear all skipped statuses that are already received."""
+        if not self._skipped.empty() and self._skipped.queue[0][0] <= self._counter:
+            _, timer = self._skipped.get()
+            timer.cancel()
+            timer.join()
+
+    def _store_skipped_counter_values(self, status_counter: CounterValue) -> None:
+        """Store all missing status counter values and start timers for them."""
+        if not self._skipped.empty() and status_counter <= self._skipped.queue[-1][0]:
+            return
```

Comment:
the skipped already contains all counters between the last properly handled status and the oldest status received since then.
Because of that, it is ensured no skipped status counter can be missing from skipped counters queue.